### ISSUE TITLE
[On Cuda] Tensor::randn fails to generate random floats for odd no: of samples

### ISSUE THREAD

**Issue by akhildevelops** [huggingface/candle#734]
I've run the sample code, with different shapes, mentioned in Readme file (https://github.com/huggingface/candle/blob/main/README.md#get-started)

```rust
use candle_core::{Device, Tensor};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let device = Device::Cpu;

    let a = Tensor::randn(0f32, 1., (4, 5), &device)?; // This will generate random values
    let b = Tensor::randn(0f32, 1., (5, 3), &device)?; // This will error out

    let c = a.matmul(&b)?;
    println!("{c}");
    Ok(())
}
```
It will error out during runtime with the message: `Cuda(Curand(CurandError(CURAND_STATUS_LENGTH_NOT_MULTIPLE)))`.

It works if I change to CPU. After digging deeper into cudarc library, found that a `pseudorandom generator` is used for generating floats and cudarand supports only [even number of output samples](https://docs.nvidia.com/cuda/curand/group__HOST.html#group__HOST_1gb9280e447ef04e1dec4611720bd0eb69) for  `pseudorandom generator`.

How can I generate odd number of samples ?

**Comment by GeauxEric** [comment]
One workaround is to index the tensor `b` and reshape:
```
#[cfg(test)]
mod tests {

    use candle_core::{Device, IndexOp, Tensor};

    #[test]
    fn it_works() -> Result<(), Box<dyn std::error::Error>> {
        let device = Device::new_cuda(0)?;
        let a = Tensor::randn(0f32, 1., (4, 5), &device)?; // This will generate random values
        let b = Tensor::randn(0f32, 1., (5, 4), &device)?; // This is OK
        let b_ = b.i((.., 0..3))?.reshape((5,3))?; // Reshape is needed
        println!("b_ shape: {:?}", b_.shape());
        let c = a.matmul(&b_)?;
        println!("{a}");
        println!("{b}");
        println!("{c}");
        Ok(())
    }
}
```
Result:
```
b_ shape: [5, 3]
[[-1.8936,  0.9397,  1.0709, -0.0135, -0.2310],
 [-0.4147, -0.1733,  0.3322,  0.1012,  0.4609],
 [-0.5161,  0.1109, -0.1549,  2.1302,  0.7367],
 [-0.3432,  1.0344,  1.1740,  0.8580,  0.8757]]
Tensor[[4, 5], f32, cuda:0]
[[-0.4003,  0.7206, -0.9397, -0.0915],
 [ 2.3167,  1.5755, -0.7861,  0.1864],
 [ 0.2528, -0.8081, -0.5926, -1.3103],
 [-1.9073,  0.8630,  0.1234, -1.7610],
 [ 0.4231, -0.7811, -0.2071, -0.2975]]
Tensor[[5, 4], f32, cuda:0]
[[ 3.1338, -0.5808,  0.4523],
 [-0.1496, -1.1131,  0.2461],
 [-3.3269,  1.1909,  0.5999],
 [ 1.5647,  0.4901, -1.2617]]
Tensor[[4, 3], f32, cuda:0]
```

Since the underlying cuRAND only supports even numbers, maybe `candle` can hide the workaround in the `randn` API.

**Comment by LaurentMazare** [comment]
Right, I've actually just merged #793 that should hopefully fix this (at the expense of generating an additional value when the number of elements is odd).

**Comment by LaurentMazare** [comment]
Closing as hopefully fixed now, but please re-open if you run into any more issues.
